### PR TITLE
Add client side raster reprojection support

### LINKED ISSUE #37: Additional map projections (SRS) for raster formats
Cesium supports several raster formats, notably:
- WMTS;
- WMS.

In ol3cesium, we are currently restricted to a custom ol3 tile image provider.
This imagery provider gets the tiles from an ol.source.TileImage in either 4326 or 
3857 projection.

Other or different projections in ol3 and Cesium are not possible.
WMS or WMTS are not handled.

### PR DESCRIPTION
**Pull request by gberaudo**
Fixes #37.

### CODE REVIEW

**Comment by gberaudo** [comment]
@klokan, your feedbacks would be welcomed.

@ahocevar, we discussed a long time ago about this one. Thanks for your guidances back then at FOSS4G Seoul. If you find some time, a second look from you would be very appreciated.

File: src/olcs/core/olimageryprovider.js [hunk 1]
```diff
@@ -247,23 +252,46 @@ olcs.core.OLImageryProvider.prototype.getTileCredits = function(x, y, level) {
...
-    return url ? Cesium.ImageryProvider.loadImage(this, url) : this.emptyCanvas_;
+  // Perform mapping of Cesium tile coordinates to ol3 tile coordinates:
+  // 1) Cesium zoom level 0 is OpenLayers zoom level 1 for EPSG:4326
+  const z_ = this.tilingScheme_ instanceof Cesium.GeographicTilingScheme ? level + 1 : level;
+  // 2) OpenLayers tile coordinates increase from bottom to top
+  const y_ = -y - 1;
+
+  const tilegrid = this.source_.getTileGridForProjection(this.projection_);
+  if (z_ < tilegrid.getMinZoom() || z_ > tilegrid.getMaxZoom()) {
+    return this.emptyCanvas_; // no data
```
Comment (wallw-bits) [inline src/olcs/core/olimageryprovider.js]: There are several places in this function which do not follow the Cesium API reference and return something other than `Promise<Image|Canvas>|undefined`. Is that intentional based on undocumented Cesium workings?

Comment (gberaudo) [reply src/olcs/core/olimageryprovider.js]: Hi @wallw-bits, you are right it works but it is incorrect. Can you come up with a PR wrapping the returned value like this: `Promise.resolve(this.emptyCanvas_)`?

File: src/olcs/core/olimageryprovider.js [hunk 2]
```diff
@@ -247,23 +252,46 @@ olcs.core.OLImageryProvider.prototype.getTileCredits = function(x, y, level) {
...
+  // not yet loaded!
+  // const image = tile.getImage();
+  // if (!image || !image.src) {
+  //   return Promise.resolve(this.emptyCanvas_); // no data
+  // }
+
+
+  const state = tile.getState();
+  if (state === ol.TileState.LOADED || state === ol.TileState.EMPTY) {
+    return tile.getImage() || undefined;
```
Comment (wallw-bits) [inline src/olcs/core/olimageryprovider.js]: This return also needs a promise wrapper if `tile.getImage()` is not undefined

Comment (gberaudo) [reply src/olcs/core/olimageryprovider.js]: Thanks, done.

**Comment by gberaudo** [comment]
I am hiding the feature behind a flag so that it can be merged. @fredj, could you please have a look to it?

**Comment by gberaudo** [comment]
This branch have issues. I am restoring it and "unmerging" the PR.

**Pull request merged**